### PR TITLE
optimization check highly available status

### LINKED ISSUE #17021: Use multiple threads to improve performance for DatabaseDiscoveryEngine.checkHighlyAvailableStatus()
Please search `// TODO query with multiple threads` in class `DatabaseDiscoveryEngine` to find the modify point.

Use multiple threads to improve performance when loading data from databases.
Please reuse `ExecutorEngine` for concurrency execute.

### PR DESCRIPTION
**Pull request by weihubeats**
Fixes #17021.

Changes proposed in this pull request:
-
-
-

### CODE REVIEW

**Review by terrymanu — CHANGES_REQUESTED** [review]
Please reuse `ExecutorEngine` for concurrency execute.

**Comment by weihubeats** [comment]
ok i used ExecutorEngine

**Review by terrymanu — APPROVED** [review]

**Review by terrymanu — CHANGES_REQUESTED** [review]
Put all data source map into one thread is not enough to improve the performance.
How about spilt each data source in each thread?

**Comment by weihubeats** [comment]
I use a separate ExecutionGroup for each task, which is multi-threaded？

**Comment by terrymanu** [comment]
> I use a separate ExecutionGroup for each task, which is multi-threaded？

No, DatabaseDiscoveryExecutorCallback should process data source only, current codes are process data source map

**Comment by weihubeats** [comment]
Thank you, I've dealt with it, is this ok?

**Comment by terrymanu** [comment]
> Thank you, I've dealt with it, is this ok?

great!

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java [hunk 1]
```diff
@@ -43,31 +49,41 @@
...
+    private static final int CPU_CORES = Runtime.getRuntime().availableProcessors();
     
     private final DatabaseDiscoveryProviderAlgorithm databaseDiscoveryProviderAlgorithm;
-    
+
     /**
      * Check environment of database cluster.
      *
-     * @param databaseName database name
+     * @param databaseName  database name
```
Comment (terrymanu) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java:60]: Please do not change original indent

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java [hunk 2]
```diff
@@ -43,31 +49,41 @@
...
      *
-     * @param databaseName database name
-     * @param groupName group name
+     * @param databaseName                  database name
+     * @param groupName                     group name
      * @param originalPrimaryDataSourceName original primary data source name
-     * @param dataSourceMap data source map
-     * @param disabledDataSourceNames disabled data source names
+     * @param dataSourceMap                 data source map
+     * @param disabledDataSourceNames       disabled data source names
```
Comment (terrymanu) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java:86]: Please do not change original indent

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java [hunk 3]
```diff
@@ -80,7 +96,7 @@ public String changePrimaryDataSource(final String databaseName, final String gr
         postReplicaDataSourceDisabledEvent(databaseName, groupName, result, dataSourceMap);
         return result;
     }
-    
+
```
Comment (terrymanu) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java:99]: Please do not change original indent

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java [hunk 4]
```diff
@@ -93,27 +109,27 @@ private Optional<String> findPrimaryDataSourceName(final Map<String, DataSource>
         }
         return Optional.empty();
     }
-    
+
```
Comment (terrymanu) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java:112]: Please do not change original indent

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java [hunk 5]
```diff
@@ -93,27 +109,27 @@ private Optional<String> findPrimaryDataSourceName(final Map<String, DataSource>
...
     private Map<String, DataSource> getActiveDataSourceMap(final Map<String, DataSource> dataSourceMap, final Collection<String> disabledDataSourceNames) {
         Map<String, DataSource> result = new HashMap<>(dataSourceMap);
         if (!disabledDataSourceNames.isEmpty()) {
             result.entrySet().removeIf(each -> disabledDataSourceNames.contains(each.getKey()));
         }
         return result;
     }
-    
+
```
Comment (terrymanu) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java:120]: Please do not change original indent

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java [hunk 6]
```diff
@@ -93,27 +109,27 @@ private Optional<String> findPrimaryDataSourceName(final Map<String, DataSource>
...
     private void postReplicaDataSourceDisabledEvent(final String databaseName, final String groupName, final String primaryDataSourceName, final Map<String, DataSource> dataSourceMap) {
         for (Entry<String, DataSource> entry : dataSourceMap.entrySet()) {
             if (!entry.getKey().equals(primaryDataSourceName)) {
                 ShardingSphereEventBus.getInstance().post(new DataSourceDisabledEvent(databaseName, groupName, entry.getKey(), createStorageNodeDataSource(loadReplicaStatus(entry.getValue()))));
             }
         }
     }
-    
+
```
Comment (terrymanu) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java:128]: Please do not change original indent

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java [hunk 7]
```diff
@@ -93,27 +109,27 @@ private Optional<String> findPrimaryDataSourceName(final Map<String, DataSource>
...
     private StorageNodeDataSource createStorageNodeDataSource(final ReplicaDataSourceStatus replicaStatus) {
         return new StorageNodeDataSource(StorageNodeRole.MEMBER, replicaStatus.isOnline() ? StorageNodeStatus.ENABLED : StorageNodeStatus.DISABLED, replicaStatus.getReplicationDelayMilliseconds());
     }
-    
+
```
Comment (terrymanu) [inline shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java:132]: Please do not change original indent

**Review by terrymanu — CHANGES_REQUESTED** [review]

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java (hunk 7 above)
Comment (weihubeats) [reply shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java]: have been processed

File: shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java (hunk 6 above)
Comment (weihubeats) [reply shardingsphere-features/shardingsphere-db-discovery/shardingsphere-db-discovery-core/src/main/java/org/apache/shardingsphere/dbdiscovery/algorithm/DatabaseDiscoveryEngine.java]: have been processed

**Comment by weihubeats** [comment]
Process all

**Review by terrymanu — APPROVED** [review]
Great job

**Pull request merged**